portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -27,17 +35,50 @@
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
-  private ArrayList<String> messages = new ArrayList<String>(Arrays.asList(
-      "Welcome to my page!", "Porfolio time!", "This is my portfolio"
-  ));
+  private int numComments = 3;
 
+  /**
+   * Get data from Datastore and return comments as a JSON file 
+   */  
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    String json = convertToJson(messages);
+    Query query = new Query("Input").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(numComments));
+    
+    ArrayList<String> comments = new ArrayList<String>();
+
+    for (Entity entity : results) {
+        comments.add((String) entity.getProperty("content"));
+    }
+
+    String json = convertToJson(comments);
     response.setContentType("application/json;");
     response.getWriter().println(json);
   }
 
+  /**
+   * Add user input from form to Datastore 
+   */  
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException  {
```

Comment:
Lots of magic constants. Use a constants file or set constants at the top of this file.